src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -633,22 +659,44 @@ private static void FlattenDeconstructVariables(ArrayBuilder<DeconstructionVaria
             return BadExpression(syntax, childNode);
         }
 
+        internal BoundDeconstructionAssignmentOperator BindDeconstruction(
+            CSharpSyntaxNode deconstruction,
+            ExpressionSyntax left,
+            ExpressionSyntax right,
+            DiagnosticBag diagnostics,
+            ref CSharpSyntaxNode declaration,
+            ref CSharpSyntaxNode expression,
+            BoundDeconstructValuePlaceholder rightPlaceholder = null)
+        {
+            DeconstructionVariable locals = BindDeconstructionVariables(left, diagnostics, ref declaration, ref expression);
+            Debug.Assert(locals.HasNestedVariables);
+            var result = BindDeconstructionAssignment(deconstruction, right, locals.NestedVariables, diagnostics, rhsPlaceholder: rightPlaceholder);
+            FreeDeconstructionVariables(locals.NestedVariables);
+            return result;
+        }
+
         /// <summary>
-        /// Returns bound variables in a tree.
-        /// For variables that are being declared, it makes locals (or fields in global statement).
-        /// If the type is unknown, a deconstruction variable pending inference is used instead (which will be replaced with a local or field later).
-        /// For expressions that don't declare variables, simply binds them and verify they are assignable to.
-        ///
+        /// Prepares locals (or fields in global statement) and lvalue expressions corresponding to the variables of the declaration.
+        /// The locals/fields/lvalues are kept in a tree which captures the nesting of variables.
+        /// Each local or field is either a simple local or field access (when its type is known) or a deconstruction variable pending inference.
         /// The caller is responsible for releasing the nested ArrayBuilders.
         /// </summary>
-        private DeconstructionVariable BindDeconstructionVariables(ExpressionSyntax node, bool isDeclaration, DiagnosticBag diagnostics)
+        private DeconstructionVariable BindDeconstructionVariables(
+            ExpressionSyntax node,
+            DiagnosticBag diagnostics,
+            ref CSharpSyntaxNode declaration,
```

Comment:
>CSharpSyntaxNode [](start = 16, length = 16)

It looks like we can use more specific type for this parameter - DeclarationExpressionSyntax.